### PR TITLE
Clean up to-device message logging

### PR DESCRIPTION
**Pull request by richvdh**
This is an attempt to clarify the log lines that get written when we build to-device messages in `share_room_key`.

Before:

```
    INFO matrix_sdk_crypto::session_manager::group_sessions: Encrypted a room key and created to-device messages, session_id: "9UI/g6wSC3FFzd3OkqT4m4hQToBOmV9WJpOIahD1ges", request_count: 2, messages: {RequestId("89ff27044bc54f17b5efe961d434851c"): {"@a:xps9320.sw1v.org": {DeviceId("VZWFLHTTWM"): MessageId("7398e9613a5d4a71a5b8b05d41422100")}}, RequestId("ae37932ce6104c9e974127201b0c6924"): {"@a:xps9320.sw1v.org": {DeviceId("FOQLOQFUSM"): MessageId("4b641b49e7534aefb78b9b45aa690840"), DeviceId("ISZBYLMAHN"): MessageId("b04d76031044454a84b749bef70de48f"), DeviceId("JPWCLBAXLN"): MessageId("29325466609f4101b8c86cb9f7544ab9"), DeviceId("LAITPBHFRV"): MessageId("acf050afbbc1420184a663f26f266c9c"), DeviceId("VYKNAUCFEU"): MessageId("02727212a8c948708dc0bec535185b69"), DeviceId("WSXTJMMFQC"): MessageId("e253a74c13b642b0b0015dff0e88ed1b")}, "@t:xps9320.sw1v.org": {DeviceId("MBQPLBQNJE"): MessageId("4a39c21d3f2a4236bdc90fa332e66f19"), DeviceId("ZTIWJHKGGA"): MessageId("96203af207a6466193bdabb80de1ec6a")}}}, withheld_messages: {}
        at /home/rav/work/matrix-rust-sdk/crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs:740
        in matrix_sdk_crypto::session_manager::group_sessions::share_room_key with room_id="!xBGdAuzcyWhesEDzge:xps9320.sw1v.org"s
```

After:

```
   INFO matrix_sdk_crypto::session_manager::group_sessions: Created batch of to-device messages of type m.room.encrypted, request_id: "b9bebc1a3e6645e1ad835288608dea66", message_list: [("e7ea9ea5bf134df6bf64edca4bf0b40c", "@richvdh:matrix.org", "AHLROBHANA"), ("0e9c9ea0a5834b9687148a73d637e95a", "@richvdh:matrix.org", "HWRQYODUFQ"), ("cf198d736f2a4665b2e0f721a51e3ca9", "@richvdh:matrix.org", "JKUAZXEPPR"), ("f44c1becadca481d954af5fbbf2686f2", "@richvdh:matrix.org", "MFKNTRGGNY"), ("6b43ec0b89b246dd8edac0f560ba9d3f", "@richvdh:matrix.org", "RDBLNSGACK"), ("b570532f26304718b1db56f3369e5edf", "@richvdh:matrix.org", "UAHXKIDIPI"), ("ef6db117060e47c1928731e7ef175167", "@richvdh:sw1v.org", "EOZDSWJVGZ"), ("103ee843abe549c48d26098467e35f51", "@richvdh:sw1v.org", "FVAOSSMDQL"), ("c86dfe23610d49a1a76aa56b1630c454", "@richvdh:sw1v.org", "FXPDLIQYHN"), ("d60b483de45240498b32947d8477c93b", "@richvdh:sw1v.org", "HKJDOCYFVC"), ("4a993d60b8be4fcc8aaaa7637a1a2c1a", "@richvdh:sw1v.org", "HSTVHNERPD"), ("cd4bf466b50f49e1b37ef6494cd05aa6", "@richvdh:sw1v.org", "SBXXDQMSKE"), ("11c70ce7cc9b4d45b5117a06430adfd4", "@richvdh:sw1v.org", "YZIQIRLUKH"), ("8cdf49901ffb4abca01cec828e73658b", "@richvdh:sw1v.org", "ZGCLWKIWBA"), ("9d5511c55c7a4c27bfdb1f355b9c406c", "@richvdh:sw1v.org", "ZOIGEATOIP")]
    at /home/rav/work/matrix-rust-sdk/crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs:659
    in matrix_sdk_crypto::session_manager::group_sessions::share_room_key with room_id="!lfeeJrGldjNZsqCPak:matrix.org" session_id="mTtGalpFy1eNnQjKunp+c+BQW8Pn58sbEfb0FYm/RFc"
```

In particular:
 * we log each batch separately
 * the session id is moved to the span, where it is also useful for other log lines
 * we don't repeat `MessageId` and `DeviceId` for each message.

### CODE REVIEW

File: crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs [hunk 1]
```diff
@@ -704,6 +697,21 @@ impl GroupSessionManager {
         result
     }
 
+    #[cfg(not(feature = "message-ids"))]
+    /// Given a to-device request, build a recipient map suitable for logging
+    ///
+    /// Returns a list of pairs of (user id, device_id)
```
Comment (jplatte) [inline crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs:703]: Docs before attributes, please.

```suggestion
    /// Given a to-device request, build a recipient map suitable for logging
    ///
    /// Returns a list of pairs of (user id, device_id)
    #[cfg(not(feature = "message-ids"))]
```

File: crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs [hunk 2]
```diff
@@ -429,12 +423,7 @@ impl GroupSessionManager {
             withheld_devices.extend(withheld_recipients);
         }
 
-        trace!(
-            should_rotate = should_rotate,
-            session_id = outbound.session_id(),
-            room_id = outbound.room_id().as_str(),
-            "Done calculating group session recipients"
-        );
+        trace!(should_rotate = should_rotate, "Done calculating group session recipients");
```
Comment (jplatte) [inline crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs:426]:
```suggestion
        trace!(should_rotate, "Done calculating group session recipients");
```

File: crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs [hunk 3]
```diff
@@ -664,86 +653,63 @@ impl GroupSessionManager {
...
-        info!(
-            session_id = session.session_id(),
-            request_count = requests.len(),
-            ?messages,
-            ?withheld_messages,
-            "Encrypted a room key and created to-device messages"
-        );
+    /// Given a to-device request, build a recipient map suitable for logging
+    ///
+    /// Returns a list of pairs of (user id, device_id)
```
Comment (jplatte) [inline crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs:702]:
```suggestion
    /// Given a to-device request, build a recipient map suitable for logging.
    ///
    /// Returns a list of pairs of (user id, device_id).
```

File: crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs [hunk 4]
```diff
@@ -664,86 +653,63 @@ impl GroupSessionManager {
...
-        #[cfg(feature = "message-ids")]
-        #[derive(Debug, Deserialize)]
-        #[serde(transparent)]
-        struct MessageId<'a>(&'a str);
-
-        #[cfg(feature = "message-ids")]
-        #[derive(Deserialize)]
+    /// Given a to-device request, build a recipient map suitable for logging
+    ///
+    /// Returns a list of triples of (message_id, user id, device_id)
```
Comment (jplatte) [inline crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs:670]:
```suggestion
    /// Given a to-device request, build a recipient map suitable for logging.
    ///
    /// Returns a list of triples of (message_id, user id, device_id).
```

**Review by jplatte — COMMENTED** [review]
To-device stuff is really not my area of expertise, so probably good to get another pair of eyes on this.

File: crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs (hunk 4 above)
Comment (richvdh) [reply crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs:670]: sorry, I always forget this.

File: crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs [hunk 5]
```diff
@@ -664,86 +653,63 @@ impl GroupSessionManager {
...
+    /// Returns a list of triples of (message_id, user id, device_id).
+    #[cfg(feature = "message-ids")]
+    fn to_device_request_to_log_list(
+        request: &Arc<ToDeviceRequest>,
+    ) -> Vec<(String, String, String)> {
+        #[derive(serde::Deserialize)]
         struct ContentStub<'a> {
-            #[serde(borrow, rename = "org.matrix.msgid")]
-            message_id: MessageId<'a>,
+            #[serde(borrow, default, rename = "org.matrix.msgid")]
```
Comment (poljar) [inline crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs:677]: Why the default and `Option` here? If we are using the `message-ids` feature we should always have the message ids.

File: crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs [hunk 6]
```diff
@@ -664,86 +653,63 @@ impl GroupSessionManager {
...
-
-        #[cfg(feature = "message-ids")]
-        #[derive(Deserialize)]
+    /// Given a to-device request, build a recipient map suitable for logging.
+    ///
+    /// Returns a list of triples of (message_id, user id, device_id).
+    #[cfg(feature = "message-ids")]
+    fn to_device_request_to_log_list(
+        request: &Arc<ToDeviceRequest>,
+    ) -> Vec<(String, String, String)> {
```
Comment (poljar) [inline crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs:674]: I feel like a custom type with the desired `Debug` implementation would have been a bit prettier but not that important.

**Review by poljar — APPROVED** [review]
Left a question/nit, looks good otherwise.

File: crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs (hunk 6 above)
Comment (richvdh) [reply crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs:674]: Noted. I'm not inclined to redo this at this point, but will bear it in mind for similar changes in the future.

**Pull request merged**